Skip updating Endpoints if no relevant fields change

tnqn opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When comparing EndpointSubsets and Endpoints, we ignore the difference
in ResourceVersion of Pod to avoid unnecessary updates caused by Pod
updates that we don't care, e.g. annotation update.

Otherwise periodic Service resync would intensively update Endpoints or
EndpointSlice whose Pods have irrelevant change between two resyncs,
leading to delay in processing newly created Services. In a scale
cluster with thousands of such Endpoints, we observed 2 minutes of
delay when the resync happens.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #108077

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Prevent unnecessary Endpoints and EndpointSlice updates caused by Pod ResourceVersion change
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

tnqn committed:
Skip updating Endpoints and EndpointSlice if no relevant fields change

When comparing EndpointSubsets and Endpoints, we ignore the difference
in ResourceVersion of Pod to avoid unnecessary updates caused by Pod
updates that we don't care, e.g. annotation update.

Otherwise periodic Service resync would intensively update Endpoints or
EndpointSlice whose Pods have irrelevant change between two resyncs,
leading to delay in processing newly created Services. In a scale
cluster with thousands of such Endpoints, we observed 2 minutes of
delay when the resync happens.

aojea commented:
but ... this way you don't publish the new resource version of the pods on the endpoints, I mean, the endpoints are not showing the latest information, is this ok? @robscott @liggitt

liggitt commented:
> it does copy

I'd be *very* interested to know why… uid I could see mattering… resourceVersion… not so much

aojea commented:
this was changed here
https://github.com/kubernetes/kubernetes/pull/91399

tnqn commented:
@aojea @liggitt @robscott thanks for your comments.
> interesting, this PR was supposed to do that https://github.com/kubernetes/kubernetes/pull/50934

#50934 skipped updating Endpoints by not triggering `syncService` when the received Pod event doesn't contain relevant change. However, `syncService` can also be triggered by periodic service resync and regular service update.

I also agree not setting resourceVersion altogether might make more sense, was just not sure if there was any reason to have it, though its value may have been stale since PR #50934. I could update the PR to remove resourceVersion if @thockin could confirm it's not needed anywhere.

aojea commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -513,8 +513,10 @@ func (e *Controller) syncService(ctx context.Context, key string) error {
 	}
+	// When comparing the subsets, we ignore the difference in ResourceVersion of Pod to avoid unnecessary Endpoints
+	// updates caused by Pod updates that we don't care, e.g. annotation update.
 	if !createEndpoints &&
-		apiequality.Semantic.DeepEqual(currentEndpoints.Subsets, subsets) &&
+		endpointutil.EndpointSubsetsEqualIgnoreResourceVersion(currentEndpoints.Subsets, subsets) &&
```
It seems that we also win on performance
```
go test -benchmem -run=^$ -bench ^BenchmarkEndpointSubsetEquality.*$ k8s.io/kubernetes/pkg/controller/endpoint -v
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/controller/endpoint
cpu: Intel(R) Xeon(R) CPU E5-2678 v3 @ 2.50GHz
BenchmarkEndpointSubsetEquality
BenchmarkEndpointSubsetEquality-24                        105562             11419 ns/op            1769 B/op         75 allocs/op
BenchmarkEndpointSubsetEqualityResourceVersion
BenchmarkEndpointSubsetEqualityResourceVersion-24         546507              2175 ns/op             192 B/op         14 allocs/op
PASS
ok      k8s.io/kubernetes/pkg/controller/endpoint       2.554s
```

```diff
diff --git a/pkg/controller/endpoint/endpoints_controller_test.go b/pkg/controller/endpoint/endpoints_controller_test.go
index a6a2fe5cf58..108bb0c83e3 100644
--- a/pkg/controller/endpoint/endpoints_controller_test.go
+++ b/pkg/controller/endpoint/endpoints_controller_test.go
@@ -27,6 +27,7 @@ import (
        "time"
 
        v1 "k8s.io/api/core/v1"
+       apiequality "k8s.io/apimachinery/pkg/api/equality"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/runtime"
        "k8s.io/apimachinery/pkg/runtime/schema"
@@ -43,6 +44,7 @@ import (
        endptspkg "k8s.io/kubernetes/pkg/api/v1/endpoints"
        api "k8s.io/kubernetes/pkg/apis/core"
        controllerpkg "k8s.io/kubernetes/pkg/controller"
+       endpointutil "k8s.io/kubernetes/pkg/controller/util/endpoint"
        utilnet "k8s.io/utils/net"
        utilpointer "k8s.io/utils/pointer"
 )
@@ -2222,6 +2224,62 @@ func TestTruncateEndpoints(t *testing.T) {
        }
 }
 
+func BenchmarkEndpointSubsetEquality(b *testing.B) {
+       s1 := []v1.EndpointSubset{{
+               Addresses: []v1.EndpointAddress{
+                       {IP: "1.2.3.4", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod0", Namespace: "ns"}},
+                       {IP: "1.2.3.5", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod1", Namespace: "ns"}},
+                       {IP: "1.2.3.6", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod2", Namespace: "ns"}},
+               },
+               Ports: []v1.EndpointPort{
+                       {Name: "port0", Port: 8080, Protocol: "TCP"},
+                       {Name: "port1", Port: 8088, Protocol: "TCP"},
+               },
+       }}
+       s2 := []v1.EndpointSubset{{
+               Addresses: []v1.EndpointAddress{
+                       {IP: "1.2.3.4", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod0", Namespace: "ns"}},
+                       {IP: "1.2.3.5", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod1", Namespace: "ns"}},
+                       {IP: "1.2.3.6", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod2", Namespace: "ns"}},
+               },
+               Ports: []v1.EndpointPort{
+                       {Name: "port0", Port: 8080, Protocol: "TCP"},
+                       {Name: "port1", Port: 8088, Protocol: "TCP"},
+               },
+       }}
+       for n := 0; n < b.N; n++ {
+               apiequality.Semantic.DeepEqual(s1, s2)
+       }
+}
+
+func BenchmarkEndpointSubsetEqualityResourceVersion(b *testing.B) {
+       s1 := []v1.EndpointSubset{{
+               Addresses: []v1.EndpointAddress{
+                       {IP: "1.2.3.4", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod0", Namespace: "ns"}},
+                       {IP: "1.2.3.5", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod1", Namespace: "ns"}},
+                       {IP: "1.2.3.6", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod2", Namespace: "ns"}},
+               },
+               Ports: []v1.EndpointPort{
+                       {Name: "port0", Port: 8080, Protocol: "TCP"},
+                       {Name: "port1", Port: 8088, Protocol: "TCP"},
+               },
+       }}
+       s2 := []v1.EndpointSubset{{
+               Addresses: []v1.EndpointAddress{
+                       {IP: "1.2.3.4", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod0", Namespace: "ns"}},
+                       {IP: "1.2.3.5", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod1", Namespace: "ns"}},
+                       {IP: "1.2.3.6", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod2", Namespace: "ns"}},
+               },
+               Ports: []v1.EndpointPort{
+                       {Name: "port0", Port: 8080, Protocol: "TCP"},
+                       {Name: "port1", Port: 8088, Protocol: "TCP"},
+               },
+       }}
+       for n := 0; n < b.N; n++ {
+               endpointutil.EndpointSubsetsEqualIgnoreResourceVersion(s1, s2)
+       }
+}
+
 func TestEndpointPortFromServicePort(t *testing.T) {
        http := utilpointer.StringPtr("http")
        testCases := map[string]struct {
```

aojea commented on pkg/controller/util/endpoint/controller_utils.go:
```diff
@@ -277,6 +277,64 @@ func (sl portsInOrder) Less(i, j int) bool {
 
+// EndpointSubsetsEqualIgnoreResourceVersion returns true if EndpointSubsets
+// have equal attributes but excludes ResourceVersion of Pod.
+func EndpointSubsetsEqualIgnoreResourceVersion(subsets1, subsets2 []v1.EndpointSubset) bool {
```
I think we should cover this function with an unit test

tnqn replied:
Thanks for the suggestion. Unit test added.

liggitt replied:
What will catch this function being incorrect if any new fields are added to EndpointSubset or any subtypes?

liggitt replied:
> Copying the subsets then resetting resourceVersions fields could but costs more, do you suggest the latter?

not really... that sounds expensive, since you'd have to deep copy first. another possibility:

Use reflection to inspect every field in the structs/substructs, modify each field one at a time in two otherwise equal objects. Make sure the equality check ignores specific fields we want to ignore (like resourceverison), and catches all other differences

tnqn replied:
Thanks for the suggestion. While I was implementating the idea, I found `Equalities` supports adding custom compare function for specific type. If we specify a function for ObjectReference, no extra overhead will be introduced. Actually it will reduce some:
```
BenchmarkEndpointSubsetEquality-40                         64736             16305 ns/op            1674 B/op         69 allocs/op
BenchmarkEndpointSubsetEqualityResourceVersion-40         421993              2920 ns/op             192 B/op         14 allocs/op
BenchmarkEndpointSubsetEqualityResourceVersion2-40         88705             14132 ns/op            1709 B/op         36 allocs/op
```

```
var SemanticIgnoreResourceVersion = conversion.EqualitiesOrDie(
	func(a, b v1.ObjectReference) bool {
		a.ResourceVersion = ""
		b.ResourceVersion = ""
		return a == b
	},
)

func EndpointSubsetsEqualIgnoreResourceVersion2(subsets1, subsets2 []v1.EndpointSubset) bool {
	return SemanticIgnoreResourceVersion.DeepEqual(subsets1, subsets2)
}
```

aojea commented:
I think that changing the function that compare the subsets is the safest options, it also improves the performance https://github.com/kubernetes/kubernetes/pull/108078#discussion_r805623556 ... and we can always revisit later the resourceVersion field  ... but since we are going to depend on the new function I think we should add unit tests https://github.com/kubernetes/kubernetes/pull/108078#discussion_r805626479

tnqn commented:
/retest

tnqn commented:
> even if we stopped setting resourceVersion (which I think we should), we'd want to avoid updating on only a resourceVersion change as well, in order to avoid a thundering herd update of every endpoint in the system on the first controller-manager restart after dropping the resourceVersion value

@liggitt Thanks for the suggestion. Then I suppose we could stop setting resourceVersion and use `EndpointSubsetsEqualIgnoreResourceVersion` in 1.24, then switch back to `apiequality.Semantic.DeepEqual` in 1.25 to avoid the overhead of copying (if we prefer not to have a custom compare function which enumerates current fields), given that most Endpoints should be updated to not include resourceVersion when running 1.24 (I confirmed restarting kubelet would trigger update of Endpoints that have Pods with probe on that Node)?

liggitt commented on pkg/controller/util/endpoint/controller_utils.go:
```diff
@@ -37,6 +38,16 @@ import (
 
+// SemanticIgnoreResourceVersion does semantic deep equality checks for objects
+// but excludes ResourceVersion of ObjectReference.
+var SemanticIgnoreResourceVersion = conversion.EqualitiesOrDie(
```
a few thoughts:
* something alone these lines might be ok, but we want to be coherent all the places we're checking endpoints / endpointslices / subsets
* we should limit use of this as a general purpose comparison... if we only expect this to be used to compare Subsets, or Endpoints, or EndpointSlices (and only for certain purposes), can we limit use to those types by exporting comparison functions just for those with appropriate doc?
* do `EndpointsEqualBeyondHash` and `objectRefPtrChanged` need to adjust here as well?

tnqn replied:
1. Added `EndpointSubsetsEqualIgnoreResourceVersion` for Endpoints API and updated `EndpointsEqualBeyondHash` for EndpointSlice API to make their behavior consistent
2.  Made the variable private and used it only in `EndpointSubsetsEqualIgnoreResourceVersion` and `EndpointsEqualBeyondHash`
3. Updated `EndpointsEqualBeyondHash` and removed `objectRefPtrChanged` which is no longer needed

aojea commented on pkg/controller/util/endpoint/controller_utils.go:
```diff
@@ -281,11 +281,11 @@ func (sl portsInOrder) Less(i, j int) bool {
 // endpoint hashing (see hashEndpoint func for more info).
 func EndpointsEqualBeyondHash(ep1, ep2 *discovery.Endpoint) bool {
-	if stringPtrChanged(ep1.NodeName, ep1.NodeName) {
+	if stringPtrChanged(ep1.NodeName, ep2.NodeName) {
```
cc: @robscott

aojea replied:
great catch :smiling_face_with_tear:

tnqn replied:
Created PR #108123 for fixing it in case separating PRs for different purposes is prefered. The commit in #108123 is also included in this PR.

tnqn commented:
/retest

tnqn commented:
@aojea @liggitt @robscott I have confirmed with @thockin, he didn't realize we had it in there either. I have updated the PR to stop publishing Pod ResourceVersion in Endpoints and EndpointSlice API and prevent Endpoints and EndpointSlice updates caused by Pod ResourceVersion change. On upgrade, the existing objects won't be updated if Pod ResourceVersion is the only difference, and the field will be set to empty when there is an actual change. Could you please take another look?

aojea commented on pkg/controller/util/endpoint/controller_utils.go:
```diff
@@ -289,40 +303,17 @@ func EndpointsEqualBeyondHash(ep1, ep2 *discovery.Endpoint) bool {
 	}
 
-	if boolPtrChanged(ep1.Conditions.Ready, ep2.Conditions.Ready) {
+	if !apiequality.Semantic.DeepEqual(ep1.Conditions, ep2.Conditions) {
```
don't we have the same problem of triggering updates with upgrades/enabling feature gate here ? 

https://github.com/kubernetes/kubernetes/blob/6acc62da7560c96f52617e744ee8be66913c72b2/pkg/controller/endpointslice/utils.go#L64-L67


https://github.com/kubernetes/kubernetes/blob/6acc62da7560c96f52617e744ee8be66913c72b2/pkg/controller/endpointslice/reconciler.go#L368-L377

https://github.com/kubernetes/kubernetes/blob/6acc62da7560c96f52617e744ee8be66913c72b2/pkg/controller/endpointslice/reconciler.go#L178-L182

tnqn replied:
Good point. EndpointSliceTerminatingCondition is enabled by default since 1.22 so it should be fine for 1.23-1.24 upgrade and I suppose we don't want to backport such change. For users that toggle the feature gate intentionally, I think they are supposed to see that the fields are updated according to their configuration?

aojea replied:
I was considering that there are some cluster that don't enable all betas too ...

aojea commented:
/lgtm
Defer approval to Rob, Tim or Jordan
/assign @robscott @thockin @liggitt

liggitt commented:
The resource version logic looks good to me, but I think we should merge the change to ignore resource version first, then separately merge the change to stop setting resource version. This makes it easier to backport just the "ignore resource version" change if we want to, and also makes sure tests still pass when the controller is encountering objects with resource version set but is ignoring it for "should I update" purposes (which is what the controller will encounter on upgrades of existing clusters)

tnqn commented:
> The resource version logic looks good to me, but I think we should merge the change to ignore resource version first, then separately merge the change to stop setting resource version. This makes it easier to backport just the "ignore resource version" change if we want to, and also makes sure tests still pass when the controller is encountering objects with resource version set but is ignoring it for "should I update" purposes (which is what the controller will encounter on upgrades of existing clusters)

@liggitt Sure, I have removed the patch that stops setting resource version and updated PR description. I will follow up with another PR after this is merged.

liggitt commented:
lgtm, would like a second from tim

thockin commented on pkg/controller/util/endpoint/controller_utils.go:
```diff
@@ -293,7 +306,20 @@ func EndpointsEqualBeyondHash(ep1, ep2 *discovery.Endpoint) bool {
+	// Otherwise all EndpointSlices in the system would be updated on the first controller-manager restart even without
+	// actual changes, leading to delay in processing legitimate updates.
+	// Its value will be set to the expected one when there is an actual change triggering update of this EndpointSlice.
+	if ep1.Conditions.Serving != nil && ep2.Conditions.Serving != nil && *ep1.Conditions.Serving != *ep2.Conditions.Serving {
```
This doesn't seem related or correct?

When Serving changes from nil to non-nil, this won't fire.  I think you want:
```
func ptrBoolEqual(a, b *bool) bool {
    if (a == nil) != (b == nil) {
        return false
    }
    if a == nil {
        return true
    }
    return *a == *b
}
```

and 

```
if ptrBoolEqual(ep1.Conditions.Serving, ep2.Conditions.Serving) {
```

liggitt replied:
> When Serving changes from nil to non-nil, this won't fire

I think that's intentional, otherwise the first controller restart with the feature enabled instantly writes all endpoints?

thockin reviewed (commented):
Why does this include Conditions checks?

thockin reviewed (commented):
Thanks!

/lgtm
/approve

tnqn commented:
@thockin Thanks for the review.
> Why does this include Conditions checks?

If it doesn't check Serving/Terminting Conditions, their changes would not trigger EndpointSlice update. It was not an issue previsouly because Pod resourceVersion also changed when any Condition changed, so `EndpointsEqualBeyondHash` would return `false` as long as it checks resourceVersion regardless of the other checks. Since we remove resourceVersion check in this PR, the function should check all fields that matter.

/retest

tnqn commented:
/retest

tnqn commented:
/test pull-kubernetes-integration
